### PR TITLE
Add IamProxyAuthProvider.

### PR DESCRIPTION
**Pull request by eap**
Note: this PR is not ready for merging and depends on (1) workbench-libs [PR 138](https://github.com/broadinstitute/workbench-libs/pull/138) being merged and (2) updated dependencies to reflect the new workbenchGoogle commit hash.

=======

This new provider uses client credentials to test permissions
on a project before granting access to a user. Thus, it proxies
the user's IAM permissions.

I, the developer opening this PR, do solemnly pinky swear that:

- [x] I've documented my API changes in Swagger

In all cases:
- [ ] Get two thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; you can delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/IamProxyAuthProvider.scala [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+import org.broadinstitute.dsde.workbench.leonardo.model.google.ClusterName
+import org.broadinstitute.dsde.workbench.leonardo.model.{LeoAuthProvider, ServiceAccountProvider}
+import org.broadinstitute.dsde.workbench.model.{UserInfo, WorkbenchEmail}
+import org.broadinstitute.dsde.workbench.model.google.{GoogleProject, IamPermission}
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class IamProxyAuthProvider(config: Config, serviceAccountProvider: ServiceAccountProvider)(implicit val system: ActorSystem) extends LeoAuthProvider(config, serviceAccountProvider) {
+
+  val applicationName: String = config.getValue("applicationName").toString()
```
Comment (rtitle) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/IamProxyAuthProvider.scala]: Nitpick: `config.getString("applicationName")` could be used.

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/IamProxyAuthProvider.scala [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+  val requiredPermissions: Set[IamPermission] = config.as[Set[String]]("requiredProjectIamPermissions").map(p => IamPermission(p))
+
+  protected def petGoogleIamDao(token: String)(implicit executionContext: ExecutionContext): GoogleIamDAO = {
+    new HttpGoogleIamDAO(applicationName, Token(() => token), "google")
+  }
+
+  protected def checkUserAccess(userInfo: UserInfo, googleProject: GoogleProject)(implicit executionContext: ExecutionContext): Future[Boolean] = {
+    val iamDAO: GoogleIamDAO = petGoogleIamDao(userInfo.accessToken.token)
+    val foundPermissions = iamDAO.testIamPermission(googleProject, requiredPermissions)
+    Future.successful(requiredPermissions == foundPermissions)
```
Comment (rtitle) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/IamProxyAuthProvider.scala]: Doesn't look right -- you're comparing a `Future[Set[IamPermission]]` with a `Set[IamPermission]`. Probably do something like this instead:
```
iamDAO.testIamPermission(googleProject, requiredPermissions).map { foundPermissions =>
   foundPermissions == requiredPermissions
}
```

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/IamProxyAuthProvider.scala [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+  /**
+    * Leo calls this method when it receives a "list clusters" API call, passing in all non-deleted clusters from the database.
+    * It should return a list of clusters that the user can see according to their authz.
+    *
+    * @param userInfo The user in question
+    * @param clusters All non-deleted clusters from the database
+    * @return         Filtered list of clusters that the user is allowed to see
+    */
+  override def filterUserVisibleClusters(userInfo: UserInfo, clusters: List[(GoogleProject, ClusterName)])(implicit executionContext: ExecutionContext): Future[List[(GoogleProject, ClusterName)]] = {
+    // Check each project for user-access exactly once, then filter by project.
```
Comment (rtitle) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/IamProxyAuthProvider.scala]: cool

**Comment by rtitle** [comment]
Looks good, some early feedback above. To make this compile you'll need to update Leo `Dependencies.scala` to reference the updated workbench-libs hash(es).

**Comment by eap** [comment]
closed this PR and replaced with [#450](https://github.com/DataBiosphere/leonardo/pull/450) based off of a local branch rather than my fork.

**Pull request closed without merging**